Make init_process_group timeout kwarg override pg_options

wconstab opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112893
* #112803
* __->__ #112611

This used to be ambiguous but the pg_options._timeout value, if passed
in, is being ignored.  Make it sane and warn if 2 values are provided.

wconstab committed:
WIP make timeout kwarg work consistently over pg_options

[ghstack-poisoned]

voznesenskym commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1284,11 +1284,16 @@ def _new_process_group_helper(
                 assert isinstance(
                     pg_options, ProcessGroupNCCL.Options
                 ), "Expected pg_options argument to be of type ProcessGroupNCCL.Options"
+                if pg_options._timeout:
+                    raise ValueError(
+                        "pg_options._timeout was specified, "
+                        "but timeout kwarg has a default value that will always override it. "
+                    )
```
Why does this option only matter under NCCL? how does this interact w/ base_pg_options._timeout?

wconstab replied:
well, gloo/mpi paths actually use this other object called 'base_pg_options' which has its _OWN_ timeout field.  That's a whole 'nother problem, which i kinda want to try fixing too.

it looks like if someone passes in a 'pg_options' object, it only gets used for nccl.  and if you use the gloo/mpi backends, the 'base_pg_options' thing gets used.  pretty sketchy.

wconstab replied:
note: if we do fix this, we might be blocked on the fact that ProcessGroupNCCL::Options is bound to python differently than ProcessGroup::Options.  that is also something to fix.

H-Huang replied:
One reason for the weirdness is that it always used to be `init_pg` -> creates 1 backend. Now `init_pg` -> can create multiple backends. However, `pg_options` in `init_pg` is only for NCCL and no other backends use it.

Hence all the other backends are using base_pg_options._timeout and this is not exposed to the user. We are just using the options as a property bag for the timeout value and backend name.

fduwjj commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1284,11 +1284,16 @@ def _new_process_group_helper(
                     pg_options, ProcessGroupNCCL.Options
                 ), "Expected pg_options argument to be of type ProcessGroupNCCL.Options"
+                if pg_options._timeout:
+                    raise ValueError(
```
do you want to just throw a waring instead of Error?

wconstab replied:
yea, good point. If this discrepancy hasn't been bothering people, then they can take their time to fix their usage.  No point introducing a bunch of breakages for people.

H-Huang replied:
I dont think many people are using timeout in pg_options, so warning sounds good. If we really wanted to we could just save the timeout and then set it again after if it has a value.

H-Huang reviewed (approved):
Okay with throwing a warning as Junjie proposed. Stamping to unblock

wconstab committed:
Update on "WIP make timeout kwarg work consistently over pg_options"




[ghstack-poisoned]

wconstab committed:
Update on "Make init_process_group timeout kwarg override pg_options"


This used to be ambiguous but the pg_options._timeout value, if passed
in, is being ignored.  Make it sane and warn if 2 values are provided.

[ghstack-poisoned]

H-Huang commented on torch/distributed/distributed_c10d.py:
```diff
@@ -1284,11 +1284,16 @@ def _new_process_group_helper(
                 assert isinstance(
                     pg_options, ProcessGroupNCCL.Options
                 ), "Expected pg_options argument to be of type ProcessGroupNCCL.Options"
+                if pg_options._timeout:
```
do we need to check if `pg_options._timeout == default timeout` instead?

wconstab committed:
Update on "Make init_process_group timeout kwarg override pg_options"


This used to be ambiguous but the pg_options._timeout value, if passed
in, is being ignored.  Make it sane and warn if 2 values are provided.

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge

wconstab commented:
@pytorchbot merge

wconstab committed:
Update on "Make init_process_group timeout kwarg override pg_options"


This used to be ambiguous but the pg_options._timeout value, if passed
in, is being ignored.  Make it sane and warn if 2 values are provided.

[ghstack-poisoned]

wconstab commented:
@pytorchbot merge
